gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -504,6 +506,173 @@ def _get_variant_calls(self, record, formats):
     return calls
 
 
+class PySamParser(VcfParser):
+  """An Iterator for processing a single VCF file using PySam.
+
+  PySam allows reading a file through either stdin stream, or through as actual
+  VCF files, for which it requires legitimate file descriptor. Since we want to
+  perform our own parallelization, we will fork our process, to use 2 pipelines
+  that will feed into PySam object - 1 will feed the data from main thread into
+  the child one, while second will get that data in child thread and feed it to
+  PySam library.
+
+  The requirement of using two pipelines comes from the design of VcfParser base
+  class - we could only use a single pipe, but it will divert the parsers.
+  """
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               file_pattern=None,  # type: str
+               representative_header_lines=None,  # type:  List[str]
+               splittable_bgzf=False,  # type: bool
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PySamParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+    self._to_child = None
+
+  def _init_child_thread(self, p_read, c_write, c_read, p_write, header_lines):
+    # Child thread's task is to populate data into the pipe that feeds
+    # VariantFile class - first by populating all of the header lines, and then
+    # by redirecting the data received from the second pipe.
+
+    os.close(p_read)
+    os.close(p_write)
+    from_parent = os.fdopen(c_read)
+    to_pysam = os.fdopen(c_write, 'w')
+    # Feed the data with a break line, since we read data line by line - if
```

Comment:
If we just use `read()`, threads would lock, as it would wait to read until EOF is reached - EOF is not reached, because processes get invoked when next line is 'requested' and wait on each other to get a new record to pass downstream. If the parser wasn't an iterator, we maybe could have just supplied all of the lines to parse, but alas, we cannot do that here.

`..next() invoked`
`......|`
`......V`
`..parent process  ---------------{raw record}---------------->  child process`
`..parent process <--{parsed record}-- VariantFile <--{raw record}--/`
`......|`
`......V`
`..return {parsed record}.`

Given that we cannot rely on EOF to do the reads, we can only rely on EOL, which we utilize via `readline()` instead. Therefore, we make sure that every line to parse has **single** newline char at the end (by trimming and then adding 1 to the end).